src/main/kotlin/com/exactpro/th2/codec/grpc/GrpcCodecService.kt

```diff
@@ -0,0 +1,123 @@
+        null
+    }
+
+    private val MessageGroupBatch.sessionAlias: String
+        get() {
+            var sessionAlias: String? = null
+            for (group in groupsList) {
+                for (message in group.messagesList) {
+                    when (sessionAlias) {
+                        null -> sessionAlias = message.sessionAlias
+                        else -> require(sessionAlias == message.sessionAlias) { "Batch contains more than one session alias" }
+                    }
+                }
+            }
+            return requireNotNull(sessionAlias) { "Batch is empty" }
+        }
+
+    override fun decode(batch: MessageGroupBatch, responseObserver: StreamObserver<MessageGroupBatch>) {
+        try {
+            val parsed = generalDecodeFunc(batch)
+            if (parsed.anyMessage(AnyMessage::hasRawMessage)) {
+                if (nextCodec == null) throw CodecException("grpc codec pipeline output contains raw messages after decoding")
+                nextCodec.decode(parsed, mapOf("session_alias" to parsed.sessionAlias), responseObserver)
+            } else {
+                responseObserver.onNext(parsed)
+                responseObserver.onCompleted()
+            }
+        } catch (t: Throwable) {
+            val errorMessage = "'decode' rpc call exception: ${t.message}"
+            eventProcessor.onErrorEvent(errorMessage, batch.messageIds, t)
+            responseObserver.onError(Status.INTERNAL.withDescription(errorMessage).withCause(t).asException())
+        }
+    }
+
+    override fun encode(batch: MessageGroupBatch, responseObserver: StreamObserver<MessageGroupBatch>) {
+        val nextCodecObserver = object : StreamObserver<MessageGroupBatch> by responseObserver {
+            override fun onNext(value: MessageGroupBatch) {
+                try {
+                    val encoded = generalEncodeFunc(value)
+
+                    if (isFirstCodecInPipeline && encoded.anyMessage(AnyMessage::hasMessage)) {
+                        throw CodecException("grpc codec pipeline output contains parsed messages after encoding")
+                    }
+
+                    responseObserver.onNext(encoded)
+                    responseObserver.onCompleted()
+                } catch (t: Throwable) {
+                    val errorMessage = "'encode' rpc call exception: ${t.message}"
+                    eventProcessor.onErrorEvent(errorMessage, batch.messageIds, t)
+                    responseObserver.onError(Status.INTERNAL.withDescription(errorMessage).withCause(t).asException())
+                }
+            }
+
+            override fun onCompleted() {}
+        }
+
+        try {
+            if (nextCodec != null && batch.anyMessage(AnyMessage::hasMessage)) {
+                nextCodec.encode(batch, mapOf("session_alias" to batch.sessionAlias), nextCodecObserver)
+            } else {
+                nextCodecObserver.onNext(batch)
+                nextCodecObserver.onCompleted()
```

Comment:
I think we have a problem here. The `onCompleted` is overridden in `nextCodecObserver` and does nothing. In this case the client will never know that we have finished processing because we never call `onCompleted` on the actual `responseObserver`

Reply:
we call `onCompleted` on the actual `responseObserver` inside overriden `onNext` method in `nextCodecObserver`.